4701-Review-asInteger-

kasperosterbye opened the pull request:
Fixes #4701

Changed the implementation to match the ansi standard.
Added a test case to match the standard and the issue raised.

kasperosterbye committed:
Fixes #4701

Ansi says to use rounded rather than truncated :
```
5.6.2.16.Message: asInteger Synopsis
Answer an integer approximating the receiver.
Definition: <number>
Answer the result of sending #rounded to the receiver.
````

Changed to `rounded`, ran all the tests that seemed relevant, and added a testcase for the issue at hand.

kasperosterbye committed:
Fixes #4701

Changed the comment to match the ansi standard

svenvc commented:
Looks OK.

I would add a bit more test cases, like 0 0.0 1/3 3/4 0.3 0.7 -1.4 -1.5 -1.75

Ducasse commented:
I think that there is an impact on the bootstrap. I restarted the build.

kasperosterbye committed:
Fixes #4701

The implementation of asInteger doing truncation instead of the ansi way of doing rounded is taken over from Squeak. It seems like the morphic system is depending on asInteger doing truncation instead of rounding. I added a comment in asInteger to warn about the fact that we are not ansi compliant.

kasperosterbye commented:
I have looked at this a bit more. I believe the right thing is to let it remain truncated rather than rounded. The reason is that Squeak is also doing truncated, and it appears to be used  for a lot of geometry calculations which computes middle lines. All of those are likely to go one off if we change this. I have reverted the implementation back, but did add a few tests to `rounded` while I was at it.

kasperosterbye commented:
@svenvc I added some extra tests to `rounded`. Actually I do not believe we need more tests for `asInteger` because that test only need to ensure we are using truncate/rounded. 
What I am trying to say, is that correct rounding is not the "subject under test". 
I am relatively new to Pharo, so I might have overlooked some principle.

frank-lesser commented:
looking at senders of asInteger changing to ANSI has quite a few side effects ( e.g Float parsing ) - for instance in Float >> #asApproximateFractionAtOrder:  asInteger has to be changed to truncated.
or (Float pi asApproximateFractionAtOrder: 3) = (355/113) answers false.  also it may afect Number parsing - so be warned this is not a trivial change

Ducasse commented:
Hi kasper
Yes I think that this is more reasonable. I like the tests :).

Stef

nicolas-cellier-aka-nice commented on src/Kernel-Tests/MagnitudeTest.class.st:
```diff
@@ -0,0 +1,8 @@
+A MagnitudeTest is a test class for testing the behavior of Magnitude
+"
+Class {
+	#name : #MagnitudeTest,
```
I see a new class without any new method... YAGNI or WIP?

kasperosterbye replied:
You seem to be better with abbreviations than I do :-).
However, you are right, and I opened issue #4730 to that matter, and hope to have a fix sometime today.
